homework10/businessinsider/task_data_parsing/task_with_async.py

```diff
@@ -0,0 +1,168 @@
+"""
+Ваша задача спарсить информацию о компаниях, находящихся в индексе S&P 500 с данного сайта:
+https://markets.businessinsider.com/index/components/s&p_500
+
+Для каждой компании собрать следующую информацию:
+
+Текущая стоимость в рублях (конвертацию производить по текущему курсу, взятому с сайта центробанка РФ)
+Код компании (справа от названия компании на странице компании)
+P/E компании (информация находится справа от графика на странице компании)
+Годовой рост/падение компании в процентах (основная таблица)
+Высчитать какую прибыль принесли бы акции компании (в процентах), если бы они были куплены на уровне 52 Week Low и
+проданы на уровне 52 Week High (справа от графика на странице компании)
+Сохранить итоговую информацию в 4 JSON файла:
+
+Топ 10 компаний с самими дорогими акциями в рублях.
+Топ 10 компаний с самым низким показателем P/E.
+Топ 10 компаний, которые показали самый высокий рост за последний год
+Топ 10 комппаний, которые принесли бы наибольшую прибыль, если бы были куплены на самом минимуме и проданы на самом
+максимуме за последний год.
+"""
+
+import asyncio
+import datetime
+import json
+import re
+from typing import Dict, List, Union
+
+import aiohttp
+from bs4 import BeautifulSoup
+
+
+async def get_current_dollar_value() -> float:
+    today = datetime.date.today().strftime("%d/%m/%Y")
+    central_bank_url = f"http://www.cbr.ru/scripts/XML_daily.asp?date_req={today}"
+    dollar_value = (
+        BeautifulSoup(await get_html_content(central_bank_url), "lxml")
+        .find("valute", id="R01235")
+        .find("value")
+        .text.replace(",", ".")
+    )
+    return float(dollar_value)
+
+
+def get_value_with_reg_exp(value, soup) -> Union[float, None]:
+    pattern = r"([0-9]*\.[0-9]*)\r\n.*" + value
+    request = re.findall(pattern, str(soup))
+    if len(request) != 0:
+        return float(request[0].replace(",", ""))
+
+
+def get_potential_profit(soup) -> Union[float, None]:
+    low52week_value = get_value_with_reg_exp(
+        "52 Week Low", soup.find("div", {"class": "snapshot"})
+    )
+    high52week_value = get_value_with_reg_exp(
+        "52 Week High", soup.find("div", {"class": "snapshot"})
+    )
+    if low52week_value and high52week_value:
+        return round((high52week_value - low52week_value) / low52week_value * 100, 2)
+
+
+async def get_html_content(url_: str) -> str:
+    async with aiohttp.ClientSession() as session:
```

Comment:
It is good practice to reuse sessions.   It takes some time to establish the connection, so then you reuse it you get some performance boost.   So let's try to do everything within a single session.

Reply:
Fixed